### PR TITLE
[CT-1040] Push Notifications (1 of 4) - Add Notification Package

### PR DESCRIPTION
**Pull request by adamfraser**
**Changelist**

Adds a new package that enables sending push notifications via Firebase

## Summary by CodeRabbit

- **New Features**
	- Introduced a notifications package with functionality for sending push notifications via Firebase.
	- Added localization support for notification messages, allowing for user-specific language preferences.
	- Implemented comprehensive configuration management for environment variables related to notifications.
	- Established a robust type system for managing various notification types and their dynamic fields.

- **Documentation**
	- Added a README file detailing setup and usage instructions for the notifications package.

- **Tests**
	- Implemented unit tests to ensure the reliability of notification message generation and sending functionality.

- **Chores**
	- Established TypeScript and ESLint configurations to maintain code quality and structure within the notifications package.

### CODE REVIEW

File: indexer/packages/notifications/src/localization.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import { LOCALIZED_MESSAGES } from './localized-messages';
+import {
+  Notification,
+  LanguageCode,
+} from './types';
+
+function replacePlaceholders(template: string, variables: Record<string, string>): string {
```
Comment (dydxwill) [inline indexer/packages/notifications/src/localization.ts:7]: can you include a docstring for what this function does?

File: indexer/packages/notifications/src/types.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+export type LanguageCode = 'en' | 'es' | 'fr' | 'de' | 'it' | 'ja' | 'ko' | 'zh';
+export function isValidLanguageCode(code: string): code is LanguageCode {
+  return ['en', 'es', 'fr', 'de', 'it', 'ja', 'ko', 'zh'].includes(code);
+}
+
+interface BaseNotification <T extends Record<string, string>> {
+  type: NotificationType,
+  titleKey: LocalizationKey,
+  bodyKey: LocalizationKey,
```
Comment (dydxwill) [inline indexer/packages/notifications/src/types.ts:52]: should we distinguish between LocalizationKeys used for title vs body?

File: indexer/packages/notifications/__tests__/localization.test.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+} from '../src/localization';
+import {
+  NotificationType,
+  NotificationDynamicFieldKey,
+  createNotification,
+  isValidLanguageCode,
+} from '../src/types';
+
+describe('deriveLocalizedNotificationMessage', () => {
+  test('should generate a correct message for DepositSuccessNotification', () => {
```
[the remark this answers is not in the document]
Comment (adamfraser) [reply indexer/packages/notifications/__tests__/localization.test.ts:12]: The compiler won't let you call the function with incorrect dynamicValues. You could get around that by explicitly ignoring the Typescript compiler but I don't think that's worth writing a test for

File: indexer/packages/notifications/src/types.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  DEPOSIT_SUCCESS_BODY = 'DEPOSIT_SUCCESS_BODY',
+  ORDER_FILLED_TITLE = 'ORDER_FILLED_TITLE',
+  ORDER_FILLED_BODY = 'ORDER_FILLED_BODY',
+  ORDER_TRIGGERED_TITLE = 'ORDER_TRIGGERED_TITLE',
+  ORDER_TRIGGERED_BODY = 'ORDER_TRIGGERED_BODY',
+}
+
+// Deeplinks for each notification
+export enum Deeplink {
+  DEPOSIT = '/profile',
```
[the remark this answers is not in the document]
Comment (adamfraser) [reply indexer/packages/notifications/src/types.ts]: Removed

File: indexer/packages/notifications/src/types.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+  switch (notificationType) {
+    case NotificationType.DEPOSIT_SUCCESS:
+      return {
+        type: NotificationType.DEPOSIT_SUCCESS,
+        titleKey: LocalizationKey.DEPOSIT_SUCCESS_TITLE,
+        bodyKey: LocalizationKey.DEPOSIT_SUCCESS_BODY,
+        topic: Topic.TRADING,
+        deeplink: Deeplink.DEPOSIT,
+        dynamicValues: dynamicValues as DepositSuccessNotification['dynamicValues'],
+      };
```
[the remark this answers is not in the document]
Comment (adamfraser) [reply indexer/packages/notifications/src/types.ts:126]: Not sure what you mean here?

File: indexer/packages/notifications/src/config.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  parseString,
+  parseSchema,
+  baseConfigSchema,
+} from '@dydxprotocol-indexer/base';
+
+export const notificationsConfigSchema = {
+  ...baseConfigSchema,
+
+  // Private Key for the Google Firebase Messaging project
+  FIREBASE_PRIVATE_KEY_BASE64: parseString({ default: 'LS0tLS1CRUdJTiBQUklWQVRFIEtFWS0tLS0tClBMQUNFSE9MREVSX0tFWV9GT1JfREVWRUxPUE1FTlQKLS0tLS1FTkQgUFJJVkFURSBLRVktLS0tLQo=' }),
```
[the remark this answers is not in the document]
Comment (adamfraser) [reply indexer/packages/notifications/src/config.ts:16]: Yep, added a comment to make that clear

File: indexer/packages/notifications/src/types.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (dydxwill) [reply indexer/packages/notifications/src/types.ts:126]:
```return {
        type: NotificationType.DEPOSIT_SUCCESS,
        titleKey: LocalizationKey.DEPOSIT_SUCCESS_TITLE,
        bodyKey: LocalizationKey.DEPOSIT_SUCCESS_BODY,
        topic: Topic.TRADING,
        deeplink: Deeplink.DEPOSIT,
        dynamicValues: dynamicValues as DepositSuccessNotification['dynamicValues'],
      } as DepositSuccessNotification;
```

**Review by dydxwill — APPROVED** [review]

**Review by dydxwill — APPROVED** [review]

**Pull request merged**